hotaru355/cuckoo-hasura
# SLR-354: http stream on yielding

## Problem

cuckoo currently provides a yielding method that returns a generator to iterate over the results of a query. Internally, however, the entire HTTP response from Hasura is loaded into memory and the method just returns a generator over that data to prevent further copies of the data being kept in memory. This is obviously not a solution for loading lot’s of data, as the processing server can run out of memory.

we would like to stream the HTTP response, parse the response into JSON _while_ it is coming in and have the `yielding` method return a generator over the incoming response.

https://dronebase.atlassian.net/browse/SLR-354

## Changes

- added the feature for `Query(Model).many(...).yielding(...)` **only**
- minor typing cleanup, small refactors
- no new tests. existing tests passing is what we want for this feature, aka keep status quo while improving memory usage

## Steps to Test or Reproduce

setup: Query 500K `asset_regions` from DEV. Stats obtained with `psutils`, see [this script](https://github.com/hotaru355/cuckoo-hasura/blob/testing/test-streaming-performance/perftesting_stream.py)

1. This branch:
```
10000
Bytes received total: 1.759 MB.
Mem used total: 6.75 MB.
20000
Bytes received total: 2.512 MB.
Mem used total: 6.75 MB.
30000
Bytes received total: 3.283 MB.
Mem used total: 6.75 MB.

...

480000
Bytes received total: 39.025 MB.
Mem used total: 7.5 MB.
490000
Bytes received total: 39.858 MB.
Mem used total: 7.5 MB.
500000
Bytes received total: 39.899 MB.
Mem used total: 7.5 MB.
======================
Total time = 36.52 s.
Bytes received total: 39.899 MB.
Mem used total: 7.5 MB.
```

2. Branch `main`
```
10000
Bytes received total: 44.427 MB.
Mem used total: 1317.164 MB.
20000
Bytes received total: 44.427 MB.
Mem used total: 1317.164 MB.
30000
Bytes received total: 44.427 MB.
Mem used total: 1317.164 MB.

...

480000
Bytes received total: 44.467 MB.
Mem used total: 1317.414 MB.
490000
Bytes received total: 44.468 MB.
Mem used total: 1317.414 MB.
500000
Bytes received total: 44.468 MB.
Mem used total: 1317.414 MB.
======================
Total time = 45.55 s.
Bytes received total: 44.468 MB.
Mem used total: 176.664 MB.
```
